chat-bots/src/Control/Monad/ListT.hs

```diff
@@ -41,44 +45,76 @@ newtype ListT m a = ListT
   }
 
 instance Functor m => Functor (ListT m) where
-  fmap :: Functor m => (a -> b) -> ListT m a -> ListT m b
+  fmap :: (a -> b) -> ListT m a -> ListT m b
   fmap f (ListT ma) = ListT $ fmap (bimap f (fmap f)) $ ma
 
 instance Monad m => Applicative (ListT m) where
-  pure :: Monad m => a -> ListT m a
-  pure = ListT . return . (`ConsF` emptyListT)
+  pure :: a -> ListT m a
+  pure a = a <$ Functor.introduce @_ @() ()
 
-  (<*>) :: Monad m => ListT m (a -> b) -> ListT m a -> ListT m b
-  (<*>) = ap
+  liftA2 :: (a -> b -> c) -> ListT m a -> ListT m b -> ListT m c
+  liftA2 f xs ys = uncurry f <$> Functor.combine (xs, ys)
 
-instance Monad m => Alternative (ListT m) where
-  empty :: Monad m => ListT m a
-  empty = emptyListT
+instance Monad m => Functor.Semigroupal (->) (,) (,) (ListT m) where
+  combine :: (ListT m x, ListT m x') -> ListT m (x, x')
+  combine = uncurry (liftM2 (,))
+
+instance Monad m => Functor.Unital (->) () () (ListT m) where
+  introduce :: () -> ListT m ()
+  introduce () = ListT $ return (() `ConsF` emptyListT)
 
-  (<|>) :: Monad m => ListT m a -> ListT m a -> ListT m a
-  ListT m <|> ListT n = ListT $ do
+instance Monad m => Alternative (ListT m) where
+  empty :: ListT m a
+  empty = absurd <$> Functor.introduce ()
+
+  (<|>) :: ListT m a -> ListT m a -> ListT m a
+  (<|>) xs ys = merge' <$> Functor.combine @_ @Either (xs, ys)
+
+instance Monad m => Functor.Semigroupal (->) Either (,) (ListT m) where
+  combine :: (ListT m x, ListT m x') -> ListT m (Either x x')
+  combine (ListT xs, ListT ys) =
+    ListT $ do
+      xs' <- xs
+      ys' <- ys
+      pure $ case (xs', ys') of
+        (NilF, NilF) -> NilF
+        (ConsF x' xs'', NilF) -> ConsF (Left x') (Left <$> xs'')
+        (NilF, ConsF y' ys'') -> ConsF (Right y') (Right <$> ys'')
+        (ConsF x' xs'', ConsF y' ys'') ->
+          ConsF (Left x') $ ListT $ pure $ ConsF (Right y') $ Functor.combine (xs'', ys'')
+
+instance Applicative m => Functor.Unital (->) Void () (ListT m) where
+  introduce :: () -> ListT m Void
+  introduce () = ListT $ pure NilF
+
+instance Monad m => Semialign (ListT m) where
+  align :: ListT m a -> ListT m b -> ListT m (These a b)
+  align = curry Functor.combine
```

Comment:
We should be careful with instances like this. We've basically established a pattern of using terms like `Alternative` or `Alignable` to refer to _simplified_ versions of existing typeclasses.

In reality these typeclasses often have additional laws or superclass constraints which aren't discharged by the laws of a mere monoidal functor. E.g. "official" `Alternative` (for whatever reason) has an `Applicative` superclass constraint, and (depending on who you ask) even requires some kind of distributivity between the two. Similarly "official" `Semialign` has a bunch of lattice-like laws. In both cases, an arbitrary monoidal functor of the appropriate type doesn't automatically satisfy these additional conditions.

It would be nice if the monoidal functor part of these typeclasses was cleanly demarcated and named in the class hierarchy, but failing that, we might have to come up with our own names.

The point of all this rambling is to give you pause before you define `Semigroupal ... => Semialign ...`. This is often the case, but is not automatically the case, because `Semialign` comes from a different package, and has laws beyond the "monoidal functor core" we've recognized.

Reply:
Ahh this makes a lot of sense and answers my earlier question about why you prefer to define Monoidal in terms of Applicative/Alternative. I'll switch over my definitions to go the other way and use the soon to be exported `From*` newtypes.